### ISSUE TITLE
[Critical] Invalid Article Purchasing Price And Discount (Whole data is incorrect?)

### ISSUE THREAD

**Issue by abdullahbaa5** [selectline-software/selectline-api#379]
### Details
- Current version number: 23.3.2.0
- Function or route (request):
GetPurchasingPrices
Articles/{articleNumber}/Purchasing/Prices/{supplierNumber} 

The data inside the selectline software:
![image](https://github.com/selectline-software/selectline-api/assets/38984510/a99c9f9e-6eb2-4958-98f1-3beb43f45fcb)

The data returned by the api:
![image](https://github.com/selectline-software/selectline-api/assets/38984510/16e6ed58-8329-45e1-867f-974850cbb124)

As you can see even if you deduct the tax from the final price, it is the wrong price, and the discount is completely missing...

The customer selling price is also wrong however the discount for customer selling price is correct.

Since customer sale price and supplier purchasing price are new, I think there are some critical bugs.

**Comment by abdullahbaa5** [comment]
This issue has halted my development for more than 3 weeks.

Is there no resolution to this?

**Comment by abdullahbaa5** [comment]
> Sorry for the late reply. We will fix this function with the next hotfix. To get the changes it is needed to reorganisate the database by the programm.
> 
> As a SelectLine Partner, please contact support directly if you have critical errors like this. They will refer you to this forum if it is not a support case.
> 
> You wrote: "I think there are some critical bugs." Please let us know what other bugs you mean.
> 
> Best regards

Currently there is a critical issue where if an article position is added to a supplier document.

For example the article is already in selectline with the following data:
- Price: 400
- Discount: 30

And since we can add article position in supplier document with custom price and discount, and example we add an article position in the supplier document with price 200 and discount 50.
The article data inside selectline will change from 400, 30 to 200, 50 too.

This is a critical issue combined with the (Get Supplier Article Price) as the supplier article price is not returning the correct price and not even returning the discount at all, and if we use those values to add to supplier document, not only the supplier document is wrong, but now the articles data in selectline is updated wrong too.

To Summarize
- Adding a article with custom price and discount to any order document updates its data in the master database.
- Get customer article price and get supplier article price is not returning the correct price and not returning discount 1 and discount 2 at all.

**Comment by MatthiasGuse** [comment]
Hello,

today, we released the Version 23.3.5 with the fix for the supplier article price.

Your Points:

- Adding a article with custom price and discount to any order document updates its data in the master database.
  - This is a feature - see "PP Update"
![grafik](https://github.com/selectline-software/selectline-api/assets/31447000/38d8a708-f342-44cf-9f43-e789c638b4b8)

- Get customer article price and get supplier article price is not returning the correct price and not returning discount 1 and discount 2 at all.
  - "supplier article price" - is corrected 
  - "customer article price" - works correct in our eyes - please give a detailed sample if you´re shure

Regards

**Comment by abdullahbaa5** [comment]
> Hello,
> 
> today, we released the Version 23.3.5 with the fix for the supplier article price.
> 
> Your Points:
> 
> * Adding a article with custom price and discount to any order document updates its data in the master database.
>   
>   * This is a feature - see "PP Update"
>     ![grafik](https://private-user-images.githubusercontent.com/31447000/324804661-38d8a708-f342-44cf-9f43-e789c638b4b8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQzNzcxNjAsIm5iZiI6MTcxNDM3Njg2MCwicGF0aCI6Ii8zMTQ0NzAwMC8zMjQ4MDQ2NjEtMzhkOGE3MDgtZjM0Mi00NGNmLTlmNDMtZTc4OWM2MzhiNGI4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDI5VDA3NDc0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZiY2E2YmVlODZhNGQzY2ZmMWI4ZTIwZjE5YWU1NTJiMmY4NDM5ZTU0MGViMTgzOWY2MmVmYzQ1ZTc3ODMyMDEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.KQM4SJ-oiHjktomqethntWZy8nWf4hyuxsrqExxwqsE)
> * Get customer article price and get supplier article price is not returning the correct price and not returning discount 1 and discount 2 at all.
>   
>   * "supplier article price" - is corrected
>   * "customer article price" - works correct in our eyes - please give a detailed sample if you´re shure
> 
> Regards

Hey!

I just got a chance to test the new version out. Though now the values are almost close, they are still not the same.

Example:
![image](https://github.com/selectline-software/selectline-api/assets/38984510/94f1d7e3-c603-490d-aab1-521c01fe8bcc)

And what the api returns:
![image](https://github.com/selectline-software/selectline-api/assets/38984510/75a2f7e8-ef16-4e82-9099-4653748bbeb1)

now if we do a little maths:
The Unit Price given is: 159.934499........

The discount given is 55% so price with discount applied will be:
![image](https://github.com/selectline-software/selectline-api/assets/38984510/bfb91149-7499-4a39-af54-029fc4d2c548)

which is the final price in the api. however, if we now remove the tax rate of 7.7% from the price:
![image](https://github.com/selectline-software/selectline-api/assets/38984510/44808cf3-a454-4120-baed-8e8af7d05f8b)
the returned price is 66.428795....... which is almost CHF 0.4 difference.

May I know if I am missing something in my calculations as a discrepancy like this in the long term can cause huge issues not only for the company but in legal finance/accounting terms too. Thank you for your quick fix.

**Comment by MatthiasGuse** [comment]
Hello,
Yes, such a manual calculation error will cause major problems in accounting.
If you add 7.7% taxes to 100 CHF, you have 107.70 CHF = 107.7%. If you want to deduct the tax again, you have to calculate from 107.7% --> 107.70 (CHF) / 107.7 (%) * 100 (%) = 100 CHF. In your example: 71.97 / 107.7 * 100 = 66.82 (with rounding difference).
I would recommend you not to make such calculations and to use the "NetPrice" parameter to get a correct net price that has also been rounded correctly.

Regards
